packages/core/src/snapshot.ts

```diff
@@ -0,0 +1,163 @@
+import { bytesToHex } from "@noble/hashes/utils"
+import * as cbor from "@ipld/dag-cbor"
+
+import * as espree from "espree"
+import * as eslintScope from "eslint-scope"
+import globals from "globals"
+
+import { Snapshot, SnapshotEffect } from "@canvas-js/interfaces"
+import { assert } from "@canvas-js/utils"
+import type { ModelSchema, IndexInit, PropertyType } from "@canvas-js/modeldb"
+
+import { Canvas, CanvasConfig } from "./Canvas.js"
+import { Contract /* SnapshotActionSchema, SnapshotModelSchema, */ } from "./types.js"
+import { EffectRecord } from "./runtime/AbstractRuntime.js"
+
+// typeguards
+export const isMergeFunction = (value: unknown): value is ModelSchema["$merge"] =>
+	typeof value === "function" || value === undefined
+export const isIndexInit = (value: unknown): value is IndexInit[] => typeof value === "string" || Array.isArray(value)
+export const isPropertyTypish = (value: unknown): value is PropertyType => typeof value === "string"
+
+// function serializeActions(contract: Contract): SnapshotActionSchema {
+// 	const result: Record<string, string> = {}
+// 	for (const [name, fn] of Object.entries(contract.actions)) {
+// 		result[name] = fn.toString()
+// 	}
+// 	return result
+// }
+
+// function serializeModels(contract: Contract): SnapshotModelSchema {
+// 	const result: SnapshotModelSchema = {}
+// 	for (const [modelName, modelSchema] of Object.entries(contract.models)) {
+// 		result[modelName] = {}
+// 		for (const [key, value] of Object.entries(modelSchema)) {
+// 			if (key === "$merge") {
+// 				assert(isMergeFunction(value))
+// 				result[modelName][key] = typeof value === "function" ? value.toString() : value
+// 			} else if (key === "$indexes") {
+// 				assert(isIndexInit(value))
+// 				result[modelName][key] = value
+// 			} else {
+// 				assert(isPropertyTypish(value))
+// 				result[modelName][key] = value
+// 			}
+// 		}
+// 	}
+// 	return result
+// }
+
+// function createContractSnapshot<T extends Contract>(config: CanvasConfig<T>) {
+// 	if (typeof config.contract === "string") {
+// 		return {
+// 			file: config.contract,
+// 			hash: hashContract(config.contract),
+// 		}
+// 	} else {
+// 		return {
+// 			actions: serializeActions(config.contract),
+// 			models: serializeModels(config.contract),
+// 			hash: hashContract(config.contract),
+// 		}
+// 	}
+// }
+
+export function checkContract(contract: Contract): boolean {
+	function hasDanglingReferences(code: string) {
+		// use latest quickjs supported version (es2023)
+		const ast = espree.parse(code, { range: true, ecmaVersion: 2023 })
+		const scopeManager = eslintScope.analyze(ast, { ecmaVersion: 2023 })
+		scopeManager.globalScope.through.forEach((ref) => {
+			if (ref.identifier.name in globals.es2023 || ref.identifier.name in globals.browser) return
+			console.log("unexpected action variable access", ref.identifier.name, "at char", ref.identifier.range?.[0])
+			return true
+		})
+		return false
+	}
```

Comment:
What happens if the contract has dangling references?

Reply:
Nothing at this point; checkContract returns null. But if we start serializing functions canonically inside the Snapshot, then we should probably emit a warning.